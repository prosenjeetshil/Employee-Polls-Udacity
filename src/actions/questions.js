const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
const SAVE_VOTE = "SAVE_VOTE";
const ADD_QUESTION = "ADD_QUESTION";

const receiveQuestions = (questions) => ({
  type: RECEIVE_QUESTIONS,
  questions,
});

const saveVote = (authedUser, qid, answer) => ({
  type: SAVE_VOTE,
  authedUser,
  qid,
  answer,
});

const addQuestion = (question) => ({
  type: ADD_QUESTION,
  question,
});

export {
  RECEIVE_QUESTIONS,
  SAVE_VOTE,
  ADD_QUESTION,
  receiveQuestions,
  saveVote,
  addQuestion,
};
