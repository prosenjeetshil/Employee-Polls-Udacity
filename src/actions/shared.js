import { 
  fetchData, 
  saveQuestion, 
  saveQuestionAnswer 
} from "../utils/api";

import { 
  createQuestion, 
  receiveUsers, 
  saveAnswer 
} from "./users";

import { 
  addQuestion, 
  receiveQuestions, 
  saveVote 
} from "./questions";

import { 
  showLoading, 
  hideLoading 
} from "react-redux-loading-bar";

const handleInitialData = () => async (dispatch) => {
  dispatch(showLoading());
  const { users, questions } = await fetchData();
  dispatch(receiveUsers(users));
  dispatch(receiveQuestions(questions));
  dispatch(hideLoading());
};

const handleSaveQuestionAnswer = (questionId, answerId) => (dispatch, getState) => {
  const { authedUser } = getState();
  dispatch(showLoading());
  dispatch(saveAnswer(authedUser, questionId, answerId));
  dispatch(saveVote(authedUser, questionId, answerId));
  dispatch(hideLoading());
  return saveQuestionAnswer(authedUser, questionId, answerId);
};

const createPoll = (optionOneText, optionTwoText) => async (dispatch, getState) => {
  const { authedUser } = getState();
  dispatch(showLoading());
  try {
    const question = await saveQuestion({
      author: authedUser,
      optionOneText,
      optionTwoText,
    });
    dispatch(createQuestion(question));
    dispatch(addQuestion(question));
  } catch (e) {
    console.error("An error occurred with [createPoll]: ", e);
  }
  dispatch(hideLoading());
};

export { 
  handleInitialData, 
  handleSaveQuestionAnswer, 
  createPoll 
};
