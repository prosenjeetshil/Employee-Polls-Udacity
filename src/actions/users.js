const RECEIVE_USERS = "RECEIVE_USERS";
const SAVE_ANSWER = "SAVE_ANSWER";
const CREATE_QUESTION = "CREATE_QUESTION";

const receiveUsers = (users) => {
return { 
  type: RECEIVE_USERS, 
  users };
};

const saveAnswer = (userId, questionId, answerId) => {
return {
  type: SAVE_ANSWER,
  authedUser: userId,
  qid: questionId,
  answer: answerId,
};
};

const createQuestion = ({ id, author }) => {
return {
  type: CREATE_QUESTION,
  id,
  author,
};
};

export { 
  RECEIVE_USERS, 
  SAVE_ANSWER, 
  CREATE_QUESTION, 
  receiveUsers, 
  saveAnswer, 
  createQuestion 
};