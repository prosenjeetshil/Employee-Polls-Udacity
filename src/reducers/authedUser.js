import { SET_AUTHED_USER } from "../actions/authedUser";

function authedUser(state = null, { type, id }) {
  switch (type) {
    case SET_AUTHED_USER:
      return id;

    default:
      return state;
  }
}

export default authedUser;
