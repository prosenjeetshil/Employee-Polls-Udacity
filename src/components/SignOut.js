import React from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";

const SignOut = ({ name, avatarURL, dispatch }) => {
  const handleSignOut = (e) => {
    e.preventDefault();
    dispatch(setAuthedUser(null));
  };

  return (
    <div className="columns is-mobile">
      <div className="column">
        <p className="title-color is-6">{name}</p>
      </div>
      <div className="column2">
        <button className="button is-normal is-light is-rounded is-responsive" onClick={handleSignOut}>
          Sign Out
        </button>
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }) => ({
  name: users[authedUser].name,
  avatarURL: users[authedUser].avatarURL
});

export default connect(mapStateToProps)(SignOut);
