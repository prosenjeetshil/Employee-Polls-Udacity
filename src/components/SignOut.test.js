import React from 'react';
import { shallow } from 'enzyme';
import SignOut from './SignOut';
import { setAuthedUser } from '../actions/authedUser';

describe('SignOut component', () => {
  let wrapper;
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.fn();
    wrapper = shallow(<SignOut name="Test User" avatarURL="" dispatch={mockDispatch} />);
  });

  it('displays the name of the user', () => {
    expect(wrapper.find('.title-color').text()).toBe('Test User');
  });

  it('dispatches the `setAuthedUser` action when the sign out button is clicked', () => {
    wrapper.find('.button').simulate('click', { preventDefault: () => {} });
    expect(mockDispatch).toHaveBeenCalledWith(setAuthedUser(null));
  });
});
