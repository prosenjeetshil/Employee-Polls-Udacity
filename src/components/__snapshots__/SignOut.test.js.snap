// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignOut should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="columns is-mobile"
      >
        <div
          class="column"
        >
          <p
            class="title-color is-6"
          >
            Sarah Edo
          </p>
        </div>
        <div
          class="column2"
        >
          <button
            class="button is-normal is-light is-rounded is-responsive"
          >
            Sign Out
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="columns is-mobile"
    >
      <div
        class="column"
      >
        <p
          class="title-color is-6"
        >
          Sarah Edo
        </p>
      </div>
      <div
        class="column2"
      >
        <button
          class="button is-normal is-light is-rounded is-responsive"
        >
          Sign Out
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
